// @electron-forge/plugin-vite/dist/VitePlugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VitePlugin = void 0;
const promises_1 = __importDefault(require("node:fs/promises"));
const node_path_1 = __importDefault(require("node:path"));
const plugin_base_1 = require("@electron-forge/plugin-base");
const debug_1 = __importDefault(require("debug"));
const vite_1 = __importDefault(require("vite"));
const ViteConfig_1 = __importDefault(require("./ViteConfig"));
const d = (0, debug_1.default)('electron-forge:plugin:vite');
class VitePlugin extends plugin_base_1.PluginBase {
    constructor() {
        super(...arguments);
        this.name = 'vite';
        this.isProd = false;
        this.watchers = [];
        this.servers = [];
        this.init = (dir) => {
            this.setDirectories(dir);
            d('hooking process events');
            process.on('exit', (_code) => this.exitHandler({ cleanup: true }));
            process.on('SIGINT', (_signal) => this.exitHandler({ exit: true }));
        };
        this.getHooks = () => {
            return {
                prePackage: [
                    (0, plugin_base_1.namedHookWithTaskFn)(async () => {
                        this.isProd = true;
                        await promises_1.default.rm(this.baseDir, { recursive: true, force: true });
                        await Promise.all([this.build(), this.buildRenderer()]);
                    }, 'Building vite bundles'),
                ],
                postStart: async (_config, child) => {
                    d('hooking electron process exit');
                    child.on('exit', () => {
                        if (child.restarted)
                            return;
                        this.exitHandler({ cleanup: true, exit: true });
                    });
                },
            };
        };
        this.startLogic = async () => {
            if (VitePlugin.alreadyStarted)
                return false;
            VitePlugin.alreadyStarted = true;
            await promises_1.default.rm(this.baseDir, { recursive: true, force: true });
            return {
                tasks: [
                    {
                        title: 'Launching dev servers for renderer process code',
                        task: async () => {
                            await this.launchRendererDevServers();
                        },
                        options: {
                            persistentOutput: true,
                            showTimer: true,
                        },
                    },
                    // The main process depends on the `server.port` of the renderer process, so the renderer process is run first.
                    {
                        title: 'Compiling main process code',
                        task: async () => {
                            await this.build(true);
                        },
                        options: {
                            showTimer: true,
                        },
                    },
                ],
                result: false,
            };
        };
        // Main process, Preload scripts and Worker process, etc.
        this.build = async (watch = false) => {
            await Promise.all((await this.configGenerator.getBuildConfig(watch)).map((userConfig) => {
                return new Promise((resolve, reject) => {
                    var _a;
                    vite_1.default
                        .build({
                        // Avoid recursive builds caused by users configuring @electron-forge/plugin-vite in Vite config file.
                        configFile: false,
                        ...userConfig,
                        plugins: [
                            {
                                name: '@electron-forge/plugin-vite:start',
                                closeBundle() {
                                    resolve();
                                    // TODO: implement hot-restart here
                                },
                            },
                            ...((_a = userConfig.plugins) !== null && _a !== void 0 ? _a : []),
                        ],
                    })
                        .then((result) => {
                        const isWatcher = (x) => typeof (x === null || x === void 0 ? void 0 : x.close) === 'function';
                        if (isWatcher(result)) {
                            this.watchers.push(result);
                        }
                        return result;
                    })
                        .catch(reject);
                });
            }));
        };
        // Renderer process
        this.buildRenderer = async () => {
            for (const userConfig of await this.configGenerator.getRendererConfig()) {
                await vite_1.default.build({
                    configFile: false,
                    ...userConfig,
                });
            }
        };
        this.launchRendererDevServers = async () => {
            var _a;
            for (const userConfig of await this.configGenerator.getRendererConfig()) {
                const viteDevServer = await vite_1.default.createServer({
                    configFile: false,
                    ...userConfig,
                });
                await viteDevServer.listen();
                viteDevServer.printUrls();
                this.servers.push(viteDevServer);
                if (viteDevServer.httpServer) {
                    // Make suee that `getDefines` in VitePlugin.ts gets the correct `server.port`. (#3198)
                    const addressInfo = viteDevServer.httpServer.address();
                    const isAddressInfo = (x) => x === null || x === void 0 ? void 0 : x.address;
                    if (isAddressInfo(addressInfo)) {
                        (_a = userConfig.server) !== null && _a !== void 0 ? _a : (userConfig.server = {});
                        userConfig.server.port = addressInfo.port;
                    }
                }
            }
        };
        this.exitHandler = (options, err) => {
            d('handling process exit with:', options);
            if (options.cleanup) {
                for (const watcher of this.watchers) {
                    d('cleaning vite watcher');
                    watcher.close();
                }
                this.watchers = [];
                for (const server of this.servers) {
                    d('cleaning http server');
                    server.close();
                }
                this.servers = [];
            }
            if (err)
                console.error(err.stack);
            // Why: This is literally what the option says to do.
            // eslint-disable-next-line no-process-exit
            if (options.exit)
                process.exit();
        };
    }
    setDirectories(dir) {
        this.projectDir = dir;
        this.baseDir = node_path_1.default.join(dir, '.vite');
    }
    get configGenerator() {
        var _a;
        return ((_a = this.configGeneratorCache) !== null && _a !== void 0 ? _a : (this.configGeneratorCache = new ViteConfig_1.default(this.config, this.projectDir, this.isProd)));
    }
}
exports.default = VitePlugin;
exports.VitePlugin = VitePlugin;
VitePlugin.alreadyStarted = false;
