// @electron-forge/plugin-vite/dist/ViteConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = __importDefault(require("node:path"));
const debug_1 = __importDefault(require("debug"));
const vite_1 = require("vite");
const plugins_1 = require("./util/plugins");
const d = (0, debug_1.default)('electron-forge:plugin:vite:viteconfig');
class ViteConfigGenerator {
    constructor(pluginConfig, projectDir, isProd) {
        this.pluginConfig = pluginConfig;
        this.projectDir = projectDir;
        this.isProd = isProd;
        this.baseDir = node_path_1.default.join(projectDir, '.vite');
        d('Config mode:', this.mode);
    }
    resolveConfig(config, configEnv = {}) {
        var _a, _b;
        // `command` is to be passed as an arguments when the user export a function in `vite.config.js`.
        // @see - https://vitejs.dev/config/#conditional-config
        (_a = configEnv.command) !== null && _a !== void 0 ? _a : (configEnv.command = this.isProd ? 'build' : 'serve');
        // `mode` affects `.env.[mode]` file loading.
        (_b = configEnv.mode) !== null && _b !== void 0 ? _b : (configEnv.mode = this.mode);
        return (0, vite_1.loadConfigFromFile)(configEnv, config);
    }
    get mode() {
        // Vite's `mode` can be passed in via command.
        // Since we are currently using the JavaScript API, we are opinionated defining two default values for mode here.
        // The `mode` set by the end user in `vite.config.js` has a higher priority.
        return this.isProd ? 'production' : 'development';
    }
    async getDefines() {
        var _a;
        const defines = {};
        const rendererConfigs = await this.getRendererConfig();
        for (const [index, userConfig] of rendererConfigs.entries()) {
            const name = this.pluginConfig.renderer[index].name;
            if (!name) {
                continue;
            }
            const NAME = name.toUpperCase().replace(/ /g, '_');
            // `server.port` is set in `launchRendererDevServers` in `VitePlugin.ts`.
            defines[`${NAME}_VITE_DEV_SERVER_URL`] = this.isProd ? undefined : JSON.stringify(`http://localhost:${(_a = userConfig === null || userConfig === void 0 ? void 0 : userConfig.server) === null || _a === void 0 ? void 0 : _a.port}`);
            defines[`${NAME}_VITE_NAME`] = JSON.stringify(name);
        }
        return defines;
    }
    async getBuildConfig(watch = false) {
        if (!Array.isArray(this.pluginConfig.build)) {
            throw new Error('"config.build" must be an Array');
        }
        const define = await this.getDefines();
        const plugins = [(0, plugins_1.externalBuiltins)()];
        const configs = this.pluginConfig.build
            .filter(({ entry, config }) => entry || config)
            .map(async ({ entry, config }) => {
            var _a;
            const defaultConfig = {
                // Ensure that each build config loads the .env file correctly.
                mode: this.mode,
                build: {
                    lib: entry
                        ? {
                            entry,
                            // Electron can only support cjs.
                            formats: ['cjs'],
                            fileName: () => '[name].js',
                        }
                        : undefined,
                    // Prevent multiple builds from interfering with each other.
                    emptyOutDir: false,
                    // 🚧 Multiple builds may conflict.
                    outDir: node_path_1.default.join(this.baseDir, 'build'),
                    watch: watch ? {} : undefined,
                },
                clearScreen: false,
                define,
                plugins,
            };
            if (config) {
                const loadResult = await this.resolveConfig(config);
                return (0, vite_1.mergeConfig)(defaultConfig, (_a = loadResult === null || loadResult === void 0 ? void 0 : loadResult.config) !== null && _a !== void 0 ? _a : {});
            }
            return defaultConfig;
        });
        return await Promise.all(configs);
    }
    async getRendererConfig() {
        var _a;
        if (!Array.isArray(this.pluginConfig.renderer)) {
            throw new Error('"config.renderer" must be an Array');
        }
        const configs = ((_a = this.rendererConfigCache) !== null && _a !== void 0 ? _a : (this.rendererConfigCache = this.pluginConfig.renderer.map(async ({ name, config }) => {
            var _a;
            const defaultConfig = {
                // Ensure that each build config loads the .env file correctly.
                mode: this.mode,
                // Make sure that Electron can be loaded into the local file using `loadFile` after packaging.
                base: './',
                build: {
                    outDir: node_path_1.default.join(this.baseDir, 'renderer', name),
                },
                clearScreen: false,
            };
            const loadResult = (_a = (await this.resolveConfig(config))) !== null && _a !== void 0 ? _a : { path: '', config: {}, dependencies: [] };
            return (0, vite_1.mergeConfig)(defaultConfig, loadResult.config);
        })));
        return await Promise.all(configs);
    }
}
exports.default = ViteConfigGenerator;
